library/src/compute/construct.rs

```diff
@@ -184,6 +186,185 @@ cast_from_value! {
+/// one of the following three ways:
+/// * If you specify the year, month and day, Typst will store it as a date internally.
+/// * If you specity the hour, minute and second, Typst will store it as a time internally.
+/// * If you specify all of year, month, day, hour, minute and second, Typst will store it as a
+/// datetime internally.
+///
+/// Depending on how it is stored, the `display` method will choose a different formatting by
+/// default.
+///
+/// ## Example
+/// ```example
+/// #datetime(year: 2012, month: 8, day: 3).display() \
+/// #datetime(hour: 2, minute: 28, second: 35).display() \
+/// #datetime(year: 2023, month: 4, day: 25, hour: 14, minute: 46, second: 59).display() \
+/// ```
+///
+/// Display: Datetime
+/// Category: construct
+/// Returns: datetime
+#[func]
+#[scope(
+    scope.define("today", today);
+    scope
+)]
+pub fn datetime(
+    /// The year of the datetime.
+    #[named]
+    year: Option<YearComponent>,
+    /// The month of the datetime.
+    #[named]
+    month: Option<MonthComponent>,
+    /// The day of the datetime.
+    #[named]
+    day: Option<DayComponent>,
+    /// The hour of the datetime.
+    #[named]
+    hour: Option<HourComponent>,
+    /// The minute of the datetime.
+    #[named]
+    minute: Option<MinuteComponent>,
+    /// The second of the datetime.
+    #[named]
+    second: Option<SecondComponent>,
+) -> Value {
+    let time = match (hour, minute, second) {
+        (Some(hour), Some(minute), Some(second)) => {
+            match time::Time::from_hms(hour.0, minute.0, second.0) {
+                Ok(time) => Some(time),
+                Err(_) => bail!(args.span, "time is invalid"),
+            }
+        }
+        _ => None,
+    };
+
+    let date = match (year, month, day) {
+        (Some(year), Some(month), Some(day)) => {
+            match time::Date::from_calendar_date(year.0, month.0, day.0) {
+                Ok(date) => Some(date),
+                Err(_) => bail!(args.span, "date is invalid"),
+            }
+        }
+        _ => None,
+    };
+
+    match (date, time) {
+        (Some(date), Some(time)) => Value::Dyn(Dynamic::new(Datetime::Datetime(
+            PrimitiveDateTime::new(date, time),
+        ))),
+        (Some(date), None) => Value::Dyn(Dynamic::new(Datetime::Date(date))),
+        (None, Some(time)) => Value::Dyn(Dynamic::new(Datetime::Time(time))),
+        (None, None) => {
+            bail!(args.span, "at least one of date or time must be fully specified")
+        }
+    }
+}
+
+struct YearComponent(i32);
+
+struct MonthComponent(Month);
+
+struct DayComponent(u8);
+
+struct HourComponent(u8);
+
+struct MinuteComponent(u8);
+
+struct SecondComponent(u8);
+
+cast_from_value!(
+    YearComponent,
+    v: i64 => match i32::try_from(v) {
+        Ok(n) => Self(n),
+        _ => Err("year is invalid")?
+    }
+);
+
+cast_from_value!(
+    MonthComponent,
+    v: i64 => match u8::try_from(v) {
```

Comment:
I tried this:

```rust
cast_from_value!(
    MonthComponent,
    v: i64 => match u8::try_from(v).and_then(|n1| Month::try_from(n1)).and_then(|n2| Self(n2)) {
        Ok(m) => m,
        _ => Err("month is invalid")?
    }
);
```

But that doesn't work because `u8::try_from` and `Month::try_from` return different error types. Not sure if there is a more elegant way to do this without making the code even longer.

Reply:
it should work by using `ok()` after each `try_from` to convert the Results to Options

Reply:
Yup!